### PR TITLE
Allow free rotation of the atlas

### PR DESCRIPTION
**Pull request by IgorTatarnikov**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
Some slices may have been taken at an angle, rotating the atlas allows viewing and registering to non-orthogonal slices of the atlas.

**What does this PR do?**
Provides a UI to rotate the atlas based on yaw, pitch and roll. This allows the user to find the correct 2D slice to register to, even if it's not aligned with the atlas axes.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality (unit & integration)
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Review by alessandrofelder — APPROVED** [review]
Thanks a lot @IgorTatarnikov and sorry for the delay in reviewing.
I made some minor comments related to legibility of the code.
I did some user testing and found UX improvements that might be worth addressing as separate issues (maybe you have already):
* users can select a sample image, and then delete the layer in napari, and then cause an index error when selecting an image in the plugin. 
* the same can happen if you delete the atlas-related layers in napari and then select a new atlas in the plugin
* if I click the rotate button in not-grid mode, it additionally toggles to grid mode - is that intended?

File: brainglobe_registration/registration_widget.py [hunk 1]
```diff
@@ -179,14 +203,38 @@ def _on_atlas_dropdown_index_changed(self, index):
...
+                self._atlas.annotation.shape[2],
+            ),
+        )
+
+        self._atlas_data_layer = self._viewer.add_image(
+            atlas_dask_array,
             name=atlas_name,
             colormap="gray",
             blending="translucent",
+            contrast_limits=[0, 350],
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_registration/registration_widget.py:229]: This was just a quick hack when I first started working on it. Do we know a good number that works for *most* atlases?

[the remark this answers is not in the document]
Comment (alessandrofelder) [reply brainglobe_registration/registration_widget.py:229]: maybe using at `0-np.percentile(central_slice, 75)` would work well enough? We can leave for now and address separately, but the hard-coded value worked badly on the bird atlas (the brain pixels were almost all saturated).

File: brainglobe_registration/registration_widget.py [hunk 2]
```diff
@@ -324,22 +385,116 @@ def _on_scale_moving_image(self, x: float, y: float):
...
+
+        # Combine the matrices. The order of operations is:
+        # 1. Translate the origin to the center of the image
+        # 2. Rotate the image
+        # 3. Translate the origin back to the top left corner
+        transform_matrix = (
+            translate_matrix
+            @ full_matrix
+            @ np.linalg.inv(post_rotate_translation)
+        )
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_registration/registration_widget.py:450]: That's a great idea, I think I over complicated it when I wrote this initially!

File: tests/test_registration_widget.py [hunk 3]
```diff
@@ -132,3 +143,62 @@ def test_scale_moving_image(
...
+
+@pytest.mark.parametrize(
+    "pitch, yaw, roll, expected_shape",
+    [
+        (0, 0, 0, (132, 80, 114)),
+        (45, 0, 0, (150, 150, 114)),
+        (0, 45, 0, (174, 80, 174)),
+        (0, 0, 45, (132, 137, 137)),
+    ],
+)
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply tests/test_registration_widget.py:156]: Regression test based on visual inspection.

**Comment by IgorTatarnikov** [comment]
Thanks for taking a look!

> I did some user testing and found UX improvements that might be worth addressing as separate issues (maybe you have already):
> 
>     * users can select a sample image, and then delete the layer in napari, and then cause an index error when selecting an image in the plugin.
> 
>     * the same can happen if you delete the atlas-related layers in napari and then select a new atlas in the plugin

This has been on my list of things to address, I'm not happy with how fragile it is at the moment. I haven't validated starting from "scratch" without re-opening `napari`. I'll make an issue.

>     * if I click the rotate button in not-grid mode, it additionally toggles to grid mode - is that intended?

This was a hacky way to reset the view, otherwise the rotated atlas doesn't quite look right. I'll come back to it and see if there's a better way of refreshing the canvas without toggling grid mode.

**Pull request merged**